Faster URIUtils::HasExtension Checks

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Replacement of  #23729 and related discussion in #23726

URIUtils::HasExtension works with lists of concatenated extensions separated by | (ex. ".mkv|.mp4|.avi|.m4v")
PR above wanted to improve performance by splitting the extensions before hand and storing them as vectors of strings, ex. { ".mkv", ".mp4", ".avi", ".m4v"}

Testing showed (see benchmarks in testing) that improved parsing of the current format of extensions lists provides performance gains similar to the use of vectors of strings, without having to change all the callers. Two parsing-improving algorithms are included in the PR.

To improve performance beyond that, std::unordered_set is needed (using hashing) but requires changes to lots of callers, all over the code base.

The current algorithm spends a lot of time lowercasing then splitting the list into a vector of strings, the new algorithms parse the string and process piece by piece.

* Two improved algorithms:
  - commit 2, straight forward, readable and C++like (2x to 20x faster than current)
  - commit 3, less readable but even faster (4x to 68x faster than current and comparable to the use of a vector of strings).
The speedup for both depends largely on the position of the matching extension in the string of extensions.

* Added many unit tests to prove the algorithms provide the same results as master, quirks included :-)

* Added a cache to the GetxxxExtensions() family of functions (ex GetVideoExtensions), which otherwise build strings to return for every call, concatenating some strings. They only need to do the concatenation once on profile load or addon addition/removal.

* Minor ~5% perf. improvement of StringUtils::EqualsNoCase / StringUtils::EndsWithNoCase if they remain (ie algo of commit 3 is chosen)

It would be possible to keep the string_view overloads of StringUtils functions added for commit 2 instead of cleaning them up.

Future: reordering the extensions in the builtin lists would provide a nice speedup as well and would allow achieving the higher range of the possible improvements for popular extensions, rather than the lower range.

copied @garbear @ksooo @howie-f as you commented on the original PR.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Noticed HasExtension() function appear in a flame graph of a library scan of movies and #23729 meant to address a similar problem with artwork.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Added unit tests covering every path in the new algorithms and clarifying corner cases of the existing code.
Kodi runs fine!

### Benchmark results

short list of 4 extensions - ".mkv|.mp4|.avi|.m4v"
| algorithm / type | debug build | release build | comment |
|:---|:---|:---|:---|
| Current master | 1x | 1x | |
| Improved algorithm 1 (commit 2) | 2x-4x | 2.5x-3.2x | depends on position of matched extension |
| Improved algorithm 2 (commit 3) | 8x | 4x-6x | depends on position of matched extension |
| vector/array of strings (not in PR) | 2.2x-3.5x | 3.x-4.3x | depends on position of matched extension |
| unordered_set (not in PR) | 9x-10x | 6.3x-6.7x | hash collisions |

long list of ~90 extensions - GetFileExtensionProvider().GetVideoExtensions()
| algorithm / type | debug build | release build | comment |
|:---|:---|:---|:---|
| Current master | 1x | 1x | |
| Improved algorithm 1 (commit 2) | 2x-20x | 1.7x-15x | depends on position of matched extension |
| Improved algorithm 2 (commit 3) | 26x-68x | 3.5x-23x | depends on position of matched extension |
| vector/array of strings (not in PR) | 2.3x-29x | 2.5x-25x | depends on position of matched extension |
| unordered_set (not in PR) | 120x | 50x-58x | rather constant hash lookup time |

The cached versions of the GetxxxExtensions() family of functions (ex GetVideoExtensions) are 60% faster than the original for debug builds. I wasn't able to measure a difference in release builds, due to testing methodology maybe, or aggressive compiler inlining/optimization...

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Overall small performance improvement, likely more noticeable on slow hardware in library scans.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

CrystalP committed:
[utils] Cache extensions in FileExtensionProvider

CrystalP committed:
[utils] Even faster algorithm for URIUtils::HasExtension()

Read each character of the list of extensions only once.

CrystalP committed:
[utils] Restore StringUtils::EqualsNoCase() and EndsWithNoCase()
remove unused string_view overloads, restore previous code

CrystalP committed:
[utils] Add doxygen for URIUtils::GetExtension

CrystalP committed:
[utils] optimization of StringUtils::EndsWihNoCase and EqualsNoCase.

EndsWihNoCase : case sensitive comparison first, similar to EqualsNoCase
EqualsNoCase: moving pointers to their next positions is not needed and wastes a bit of time if the pointed characters are not equal.

ksooo commented on xbmc/settings/AdvancedSettings.h:
```diff
@@ -100,6 +103,8 @@ struct RefreshVideoLatency
 
 typedef std::vector<TVShowRegexp> SETTINGS_TVSHOWLIST;
 
+typedef std::function<void()> AdvancedSettingsCallback;
```
We prefer `using`, not `typedef`.

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#127-type-aliases

ksooo commented on xbmc/settings/AdvancedSettings.h:
```diff
@@ -116,6 +121,9 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     bool Load(const CProfileManager &profileManager);
 
+    int RegisterSettingsLoadedCallback(AdvancedSettingsCallback callback);
+    void UnregisterSettingsLoadedCallback(int);
```
Have you intentionally not provided a name for the parameter?

CrystalP replied:
no, I missed it. Added doxygen while I was at it.

ksooo commented on xbmc/settings/AdvancedSettings.h:
```diff
@@ -386,4 +394,7 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     void Clear();
     void SetExtraArtwork(const TiXmlElement* arttypes, std::vector<std::string>& artworkMap);
+
+    mutable CCriticalSection m_critSection; //!< protects the callbacks list access
```
If this mutex is only for protecting the callbacks list, like the comment suggests, you probably should name the member accordingly, then the comment would even not be needed anymore.

CrystalP replied:
Yes, done. Access to a lot of class members should probably be protected but that would be best done as a dedicated PR. Rarely if ever is a problem because settings are typically loaded once on start and there is no uncontrolled activity. Profile switch may be more risky.

ksooo commented on xbmc/utils/FileExtensionProvider.cpp:
```diff
@@ -48,53 +52,77 @@ CFileExtensionProvider::~CFileExtensionProvider()
-  extensions += '|' + GetAddonExtensions(AddonType::AUDIODECODER);
+  if (!m_musicExtensions)
+  {
+
```
remove the blank line

ksooo commented on xbmc/utils/FileExtensionProvider.cpp:
```diff
@@ -48,53 +52,77 @@ CFileExtensionProvider::~CFileExtensionProvider()
-  extensions += '|' + GetAddonExtensions(AddonType::IMAGEDECODER);
+  if (!m_pictureExtensions)
+  {
+
```
remove the blank line

ksooo commented on xbmc/utils/FileExtensionProvider.cpp:
```diff
@@ -207,6 +235,10 @@ void CFileExtensionProvider::SetAddonExtensions(AddonType type)
           fileFolderExtensions.push_back(ext.first);
       }
     }
+    // Invalidate dependent cached extensions lists
+    m_musicExtensions = {};
+    m_pictureExtensions = {};
+    m_fileFolderExtensions = {};
```
`m_subtitlesExtensions` is missing here?!

ksooo replied:
What about `m_discStubExtensions`?

CrystalP replied:
same answer as other similar code - those 2 don't depend on addons.

ksooo commented on xbmc/utils/FileExtensionProvider.cpp:
```diff
@@ -230,6 +262,11 @@ void CFileExtensionProvider::SetAddonExtensions(AddonType type)
         }
       }
     }
+    // Invalidate dependent cached extensions lists
+    m_musicExtensions = {};
+    m_pictureExtensions = {};
+    m_subtitlesExtensions = {};
+    m_videoExtensions = {};
```
What about `m_discStubExtensions`?

CrystalP replied:
its value doesn't depend on addons.

ksooo commented on xbmc/utils/FileExtensionProvider.h:
```diff
@@ -77,15 +78,25 @@ class CFileExtensionProvider
   // Construction properties
   std::shared_ptr<CAdvancedSettings> m_advancedSettings;
   ADDON::CAddonMgr &m_addonManager;
+  int m_callbackId;
```
default init the member

CrystalP replied:
It's initialized in the constructor with the other members.

ksooo commented on xbmc/utils/StringUtils.h:
```diff
@@ -104,9 +104,9 @@ class StringUtils
-  static bool EqualsNoCase(const std::string &str1, const std::string &str2);
-  static bool EqualsNoCase(const std::string &str1, const char *s2);
-  static bool EqualsNoCase(const char *s1, const char *s2);
+  static bool EqualsNoCase(const std::string& str1, const std::string& str2);
+  static bool EqualsNoCase(const std::string& str1, const char* s2);
+  static bool EqualsNoCase(const char* s1, const char* s2);
```
Why do you only fix these formatting issues, not all/none still contained in this header? Change seems completely random.

ksooo commented on xbmc/utils/StringUtils.h:
```diff
@@ -141,8 +141,8 @@ class StringUtils
   static bool EndsWith(const std::string &str1, const char *s2);
-  static bool EndsWithNoCase(const std::string &str1, const std::string &str2);
-  static bool EndsWithNoCase(const std::string &str1, const char *s2);
+  static bool EndsWithNoCase(const std::string& str1, const std::string& str2);
+  static bool EndsWithNoCase(const std::string& str1, const char* s2);
```
Why do you only fix these formatting issues, not all/none still contained in this header? Change seems completely random.

CrystalP replied:
That's an unintended/unavoidable byproduct of trying to include the choice of 2 algorithms in the PR, one of which needs StringUtils changes, and the other one doesn't (and reverts the prev. StringUtils changes - but reverting to original generates formatting warnings).
If algo 1 / commit 2 is chosen, then the .h and .cpp format changes will match.
If the algorithm 2 / commit 3 is chosen, I'll remove the StringUtils changes+revert and the header won't change at all.

ksooo commented on xbmc/utils/URIUtils.h:
```diff
@@ -67,8 +82,8 @@ class URIUtils
    */
-  static bool HasExtension(const std::string& strFileName, const std::string& strExtensions);
-  static bool HasExtension(const CURL& url, const std::string& strExtensions);
+  static bool HasExtension(const std::string& strFileName, std::string_view strExtensions);
+  static bool HasExtension(const CURL& url, std::string_view strExtensions);
```
Could you please remove the type prefix from the string params?

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -88,35 +88,80 @@ bool URIUtils::HasExtension(const std::string& strFileName)
 }
 
-bool URIUtils::HasExtension(const CURL& url, const std::string& strExtensions)
+bool URIUtils::HasExtension(const CURL& url, std::string_view strExtensions)
```
Could you please remove the type prefix from the string param?

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -88,35 +88,80 @@ bool URIUtils::HasExtension(const std::string& strFileName)
+namespace
 {
-  if (IsURL(strFileName))
+bool IsExtensionInList(std::string_view extension, std::string_view extList)
```
Could you please remove the type prefix from the string param?

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods

ksooo commented on xbmc/utils/test/TestStringUtils.cpp:
```diff
@@ -134,6 +134,15 @@ TEST(TestStringUtils, EqualsNoCase)
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, "ta"));
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, ""));
+
+  std::string empstr = "";
```
No need for the `= ""`. default ctor is sufficient.

CrystalP committed:
[utils] Add StringUtils and URIUtils tests

CrystalP committed:
[utils] Faster algorithm for URIUtils::HasExtension()

Avoid the costly upfront string split with many string allocations and systematic lower casing.
Instead do a "lazy" split to compare segments of the list of extensions.

CrystalP commented:
Jenkins build this please

Thanks for the review, comments addressed or replied to.
Added a few unit tests to the first commit.

garbear commented:
Wanted for Beta 2?
